lib/protocol/protocol.go

```diff
@@ -823,26 +753,63 @@ func (c *rawConnection) writeUncompressedMessage(msg message) error {
 
 	totSize := 2 + hdrSize + 4 + size
 	buf := BufferPool.Get(totSize)
+	defer BufferPool.Put(buf)
 
 	// Header length
 	binary.BigEndian.PutUint16(buf, uint16(hdrSize))
 	// Header
 	if _, err := hdr.MarshalTo(buf[2:]); err != nil {
-		BufferPool.Put(buf)
 		return errors.Wrap(err, "marshalling header")
 	}
 	// Message length
 	binary.BigEndian.PutUint32(buf[2+hdrSize:], uint32(size))
 	// Message
 	if _, err := msg.MarshalTo(buf[2+hdrSize+4:]); err != nil {
-		BufferPool.Put(buf)
 		return errors.Wrap(err, "marshalling message")
 	}
 
+	var compressedSize int
+	if c.shouldCompressMessage(msg) {
+		chdr := Header{
+			Type:        c.typeOf(msg),
+			Compression: MessageCompressionLZ4,
+		}
+		chdrSize := chdr.ProtoSize()
+		if chdrSize > 1<<16-1 {
+			panic("impossibly large header")
+		}
+
+		overhead := 2 + chdrSize + 4
+		maxCompressed := overhead + lz4.CompressBound(size)
+		cbuf := BufferPool.Get(maxCompressed)
+		defer BufferPool.Put(cbuf)
+
+		compressedSize, err = lz4Compress(buf[2+hdrSize+4:], cbuf[overhead:])
+		if err != nil || compressedSize+overhead >= totSize {
+			goto write
+		}
+
+		buf, hdr, hdrSize = cbuf, chdr, chdrSize
+		totSize = compressedSize + overhead
+
+		// Header length
+		binary.BigEndian.PutUint16(buf, uint16(hdrSize))
+		// Header
+		if _, err := hdr.MarshalTo(buf[2:]); err != nil {
+			return errors.Wrap(err, "marshalling header")
+		}
+		// Message length
+		binary.BigEndian.PutUint32(buf[2+chdrSize:], uint32(compressedSize))
+	}
```

Comment:
I think it would be simpler/cleaner if this code would stay in a separate function, e.g. `compressedMessage(msg message, uncompressedSize int) ([]byte, bool, error)` that returns false if compressing increases the size. Obviously it's entirely possible I am missing something that makes this annoying/ugly.

Reply:
That means messages may have to be marshalled three times: once to know what needs to be compressed, once compressed, then again to send the uncompressed message.

I can try to separate the compression stuff into a separate method, though.

Reply:
Please see the updated PR.